09-user-feedback/client.py

```diff
@@ -16,17 +18,20 @@ def __str__(self):
         return self.content
 
 
-class _SpanIDCapturingList(list):
-    """Defers capture of the OpenTelemetry span ID until it is iterated."""
+# We need the OpenTelemetry span ID created automatically by OpenInference
+# instrumentation. Since this instrumentation applies at a low level, the
+# `request` function, we need to patch some function called inside that.
+span_id_var = contextvars.ContextVar("span_id", default=0)
+original_build_request = BaseClient._build_request
 
-    def __init__(self, iterable):
-        super().__init__(iterable)
-        self.span_id = 0
 
-    def __iter__(self):
-        ctx = trace.get_current_span().get_span_context()
-        self.span_id = ctx.span_id
-        return super().__iter__()
+def span_id_capturing_build_request(self, *args, **kwargs):
+    ctx = trace.get_current_span().get_span_context()
+    span_id_var.set(ctx.span_id)
```

Comment:
Oh I realized this is relying on context leak. Technically a `.set` is always supposed to look something like this (which would cause the value to get cleared before we can read it)

```python
token = span_id_var.set(ctx.span_id)
try:
  return original_build_request(self, *args, **kwargs)
finally:
  span_id_var.reset(token)
```

I pushed a change that should be safer, what do you think?